docs/architecture/psa-thread-safety.md

```diff
@@ -0,0 +1,275 @@
+
+Note that this does not protect access to the RNG via `mbedtls_psa_get_random`, which is guaranteed to be thread-safe when `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` is disabled.
+
+This approach is conceptually simple, but requires extra instrumentation to every function and has bad performance in a multithreaded environment since a slow operation in one thread blocks unrelated operations on other threads.
+
+## Global lock excluding slot content
+
+Have a single mutex protecting all accesses to the key store and other global variables, except that it's ok to access the content of a key slot without taking the lock if one of the following conditions holds:
+
+* The key slot is in a state that guarantees that the thread has exclusive access.
+* The key slot is in a state that guarantees that no other thread can modify the slot content, and the accessing thread is only reading the slot.
+
+Note that a thread must hold the global mutex when it reads or changes a slot's state.
+
+### Slot states
+
+For concurrency purposes, a slot can be in one of three states:
+
+* UNUSED: no thread is currently accessing the slot. It may be occupied by a volatile key or a cached key.
+* WRITING: a thread has exclusive access to the slot. This can only happen in specific circumstances as detailed below.
+* READING: any thread may read from the slot.
+
+A high-level view of state transitions:
+
+* `psa_get_empty_key_slot`: UNUSED → WRITING.
+* `psa_get_and_lock_key_slot_in_memory`: UNUSED or READING → READING. This function only accepts slots in the UNUSED or READING state. A slot with the correct id but in the WRITING state is considered free.
+* `psa_unlock_key_slot`: READING → UNUSED or READING.
+* `psa_finish_key_creation`: WRITING → READING.
+* `psa_fail_key_creation`: WRITING → UNUSED.
+* `psa_wipe_key_slot`: any → UNUSED. If the slot is READING or WRITING on entry, this function must wait until the writer or all readers have finished. (By the way, the WRITING state is possible if `mbedtls_psa_crypto_free` is called while a key creation is in progress.) See [“Destruction of a key in use”](#destruction of a key in use).
+
+The current `state->lock_count` corresponds to the difference between UNUSED and READING: a slot is in use iff its lock count is nonzero, so `lock_count == 0` corresponds to UNUSED and `lock_count != 0` corresponds to READING.
+
+There is currently no indication of when a slot is in the WRITING state. This only happens between a call to `psa_start_key_creation` and a call to one of `psa_finish_key_creation` or `psa_fail_key_creation`. This new state can be conveyed by a new boolean flag, or by setting `lock_count` to `~0`.
+
+### Destruction of a key in use
+
+Problem: a key slot is destroyed (by `psa_wipe_key_slot`) while it's in use (READING or WRITING).
+
+TODO: how do we ensure that? This needs something more sophisticated than mutexes (concurrency number >2)! Even a per-slot mutex isn't enough (we'd need a reader-writer lock).
```

Comment:
It is not very elegant but we could do it with mutexes only:
If the slot is in use when `psa_wipe_key_slot` is called, the slot is only logically deleted (a boolean field per slot). The actual wiping happens when the last reader stops reading (`lock_count` reaches 0). When a slot is logically deleted, new readers can't start reading (`lock_count` can't increase). Currently this would mean that the last `psa_unlock_key_slot` clears the slot metadata (the key material can be deleted by `psa_wipe_key_slot` as operations have a copy if they need it in later stages).
What am I missing?

Reply:
How should `psa_destroy_key` work in a single threaded scenario?
1) Return an error when the key is in use (this is what we are doing now so I presume this is not the correct answer)
2) Do a logical delete as suggested above (this doesn't guarantee that upon return the key material is wiped so this is probably not what we want either)
3) Refactor everything to do away with the copies of the key material so that the only copy is in the slot. `psa_destroy_key` deletes the only copy of the key from the slot. It is the callers responsibility to make sure that there are no ongoing operations using that key.
4) Something else

Reply:
Or since the documentation of  `psa_destroy_key()` says:
> If a key is currently in use in a multipart operation, then destroying the
key will cause the multipart operation to fail.

is it better to simply just kill any potential ongoing operation involving the key?

Reply:
> is it better to simply just kill any potential ongoing operation involving the key?

Well, yes, but you can't just kill a thread that's running a long computation, you have to wait until it finishes. It might be in the middle of talking to hardware, or have allocated blocks in local variables, etc. You have to wait until the thread reaches a point where it goes and notices that something's up. Which, in practice, means when it returns from the ongoing API call.